### PR TITLE
resolve #107 IO module out outLn etc

### LINKED ISSUE #107: Implement 'out', 'outLn', 'err' and 'errLn' functions for outputting 'Text' to corresponding handlers
I believe, the title is pretty clear. What's needed specifically:

- [ ] Create `Iris.IO` module
- [ ] Implement all four functions in this module with documentation and examples
- [ ] Reexport this module from `Iris`

### PR DESCRIPTION
**Pull request by martinhelmer**
[#107] Implement 'out', 'outLn', 'err' and 'errLn' functions for outputting 'Text' to corresponding handlers

functions added in IO module
refactored existing calls to use new functions

### Additional tasks

- [x] Documentation for changes provided/changed
- [x] Tests added
- [x] Updated CHANGELOG.md

### CODE REVIEW

File: src/Iris/IO.hs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+No linefeed at the end.
+
+@
+ghci> do Iris.out "foo" >> Iris.out "bar"
+foobarghci>
+
+@
+@since x.x.x.x
+-}
+out :: Text -> IO ()
```
Comment (chshersh) [inline src/Iris/IO.hs:53]: Let's actually make this function work in a polymoprhic `MonadIO`, so people won't need to write `liftIO`. And you actually remove `liftIO $ ...` in all places where you use `out`, `outLn` and etc.

```suggestion
out :: MonadIO m => Text -> m ()
```

File: src/Iris/IO.hs [hunk 2]
```diff
@@ -0,0 +1,95 @@
+{- |
+Module                  : Iris.IO
+Copyright               : (c) 2022 Dmitrii Kovanikov
```
Comment (chshersh) [inline src/Iris/IO.hs:3]: I guess this was copy-pasted from another module, so we can update the year 😅 
```suggestion
Copyright               : (c) 2023 Dmitrii Kovanikov
```

File: test/Test/Iris/IO.hs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+-- Silently has side effect: writes file to tmp (or current dir) and then deletes it.
+
+import System.IO (stderr, stdout)
+import System.IO.Silently (hCapture_, hSilence)
+
+import qualified Iris.IO as IO (err, errLn, out, outLn)
+
+checkStdErr :: IO a -> IO String
+checkStdErr = hCapture_ [stderr] . hSilence [stdout]
```
Comment (chshersh) [inline test/Test/Iris/IO.hs:14]: I think the `hSilence` part should be removed:

```suggestion
checkStdErr = hCapture_ [stderr]
```

Otherwise, the test `it "does not write to stderr " $ checkStdErr (IO.out "TEXT") >>= flip shouldBe ""` doesn't really check that the function doesn't output anything to `stderr`. Because `stderr` was explicitly silenced.

**Review by chshersh — CHANGES_REQUESTED** [review]
Looks neat 👍🏻 

A few suggestions to make the API more convenient, and also some improvements to tests.

File: test/Test/Iris/IO.hs (hunk 3 above)
Comment (martinhelmer) [reply test/Test/Iris/IO.hs:14]: we need to silence the handle we're not checking. otherwise erroneous behaviour will cause "TEXT" to leak into the console, messing up the test progress report. 

I believe the implementation **is** correct as it is. checkStdErr does not silence stderr. it silences stdout. 

i can also confirm that the tests **do** seem to work; hacking IO.out to print to both stdout and stderr gives ths following result:
```
   IO
    out
      writes to stdout, no LF  [✔]
      does not write to stderr  [✘]
    outLn
      writes to stdout, LF  [✔]
      does not write to stderr  [✔]
    err
      writes to sterr, no LF  [✔]
      does not write to stdout  [✔]
    errLn
      writes to stderr, LF  [✔]
      does not write to stdout  [✔]

Failures:

  test/Test/Iris/IO.hs:25:83:
  1) Iris.IO.out does not write to stderr
       expected: ""
        but got: "TEXT"
```

but maybe i'm still missing something...

Comment (chshersh) [reply test/Test/Iris/IO.hs:14]: I see now. I was confused by `silence` but I can see that both functions `checkStdOut` and `checkStdErr` are used in a pair to test the output to both `stdout` and `stderr`. A single test is not enough to make sure that the function doesn't print somewhere else, but two tests together make everything work perfectly 👍🏻 

No need to change these helper functions 😌

File: src/Iris/Interactive/Question.hs [hunk 4]
```diff
@@ -107,12 +108,12 @@ yesno question defaultAnswer = do
...
+        IO.out $ question <> " (yes/no) "
         hFlush stdout
         input <- Text.getLine
         case parseYesNo input of
             Just answer -> pure answer
             Nothing -> do
-                Text.putStrLn $ "I don't understand your answer: '" <> input <> "'"
-                Text.putStrLn "Please, answer yes or no (or y, or n)"
+                IO.out $ "I don't understand your answer: '" <> input <> "'"
+                IO.out "Please, answer yes or no (or y, or n)"
```
Comment (chshersh) [inline src/Iris/Interactive/Question.hs:118]: This should be `outLn` and not just `out`.

Ideally, if we could fully test this function, the tests could've caught this error. But I don't see a good way to test it. I was thinking about mocking `stdin` somehow to provide input. But this looks impossible (or maybe I just don't know how to do this). Tricks like `silently` allow to capture only `stdout` and `stderr` but they don't allow to substitute `stdin` which is a pity.

A proper test for this function would be to write a separate executable and use some scripting-scenario framework where you can specify input and expected output. But this is a lot of work for this small test, but maybe we'll add it later eventually anyway when things stabilise.

> P.S. I remember already writing this comment but I don't see it in the review. Maybe I forgot to submit it 🤔

File: test/Test/Iris/IO.hs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    describe "IO" $ do
+        describe "out" $ do
+            -- we need `flip` to get the expectation reported correctly in case of a failure
+            it "writes to stdout, no LF " $ checkStdOut (IO.out "TEXT") `shouldReturn` "TEXT"
+            it "does not write to stderr " $ checkStdErr (IO.out "TEXT") `shouldReturn` ""
+        describe "outLn" $ do
+            it "writes to stdout, LF " $ checkStdOut (IO.outLn "TEXT") `shouldReturn` "TEXT\n"
+            it "does not write to stderr " $ checkStdErr (IO.outLn "TEXT") `shouldReturn` ""
+        describe "err" $ do
+            it "writes to sterr, no LF " $ checkStdErr (IO.err "TEXT") `shouldReturn` "TEXT"
```
Comment (chshersh) [inline test/Test/Iris/IO.hs:30]: Typo
```suggestion
            it "writes to stderr, no LF " $ checkStdErr (IO.err "TEXT") `shouldReturn` "TEXT"
```

**Review by chshersh — CHANGES_REQUESTED** [review]
Looks great! A few minor nitpicks but this is almost done 🏁

File: src/Iris/Interactive/Question.hs (hunk 4 above)
Comment (martinhelmer) [reply src/Iris/Interactive/Question.hs:118]: Have you considered a future improvement which would be "take interactive input from file instead of stdin"? (Useful for batch installation scripts etc)

If we were to support such an improvement we'd need to store the input handle in the environment. And that would also allow for unit testing this part easier. 

Thoughts?

Comment (chshersh) [reply src/Iris/Interactive/Question.hs:118]: I wrote multiple CLI tools but I never had the need to test in batch mode from a file 🤔 

You don't need to mock `stdin` in such cases because you can always pipe the content of the file into your program:

```haskell
cat my-commands | my-program
```

For full E2E testing of CLI tools, one can use an external solution like this one:

* https://github.com/simonmichael/shelltestrunner

I can see how storing `stdin` in the environment can make testing easier but it can also make the architecture of a CLI app more complicated, confusing and potentially introduce errors when you read not from the environment `stdin` but from the global one

So I propose just changing `IO.out` to `IO.outLn` here and invest if full integration testing later.
